### PR TITLE
logging: support emitting MIPI Sys-T catalog messages

### PR DESCRIPTION
**Pull request by dcpleung**
MIPI Sys-T catalog messages are similar to dictionary logging where an ID is emitted instead of the format string. This allows the format strings to be removed from the final binary to save a few bytes. This adds the necessary bits to determine to emit catalog messages when appropriate.

See individual commits for details.

### CODE REVIEW

**Comment by dcpleung** [comment]
Updated `west.yml` as https://github.com/zephyrproject-rtos/mipi-sys-t/pull/6 has been merged.

**Review by nordic-krch — COMMENTED** [review]
Started to look into it. Wasn't yet involved much in dictionary/catalog logging.
I have 2 question:
- why string from rodata are not cataloged but appended to the log message?
- how to parse catalog mipi syst sample? I run the sample and wanted to run `log_parser.py` with xml file as database but got error.

**Comment by dcpleung** [comment]
> * why string from rodata are not cataloged but appended to the log message?

Sys-T does not use catalog for string arguments. There is no way to tag whether a string in the argument list is a string in catalog or a full string. Therefore, each string argument needs to be replaced with the full string.

> * how to parse catalog mipi syst sample? I run the sample and wanted to run `log_parser.py` with xml file as database but got error.

There is a `systprint` utility in `modules/debug/mipi-sys-t/` that needs to be built. For example, after saving the content of a qemu_x86 run in a text file, you can do `systprint -p -c <build_dir>/zephyr/mipi_syst_collateral.xml <qemu output>` to get the messages. Note that the utility looks for lines (strictly) beginning with `SYS-T RAW DATA:` for messages to be parsed. So depending on the QEMU output, the first line may not print.

File: include/logging/log_msg2.h [hunk 1]
```diff
@@ -195,19 +195,20 @@ enum z_log_msg2_mode {
 #define Z_LOG_MSG2_STACK_CREATE(_cstr_cnt, _domain_id, _source, _level, _data, _dlen, ...) \
 do { \
 	int _plen; \
+	uint32_t flags = CBPRINTF_PACKAGE_FIRST_RO_STR_CNT(_cstr_cnt) | \
```
[the remark this answers is not in the document]
Comment (dcpleung) [reply include/logging/log_msg2.h]: Updated and replaced it with a macro.

File: include/logging/log_msg2.h [hunk 2]
```diff
@@ -221,8 +222,7 @@ do { \
 #ifdef CONFIG_LOG_SPEED
 #define Z_LOG_MSG2_SIMPLE_CREATE(_cstr_cnt, _domain_id, _source, _level, ...) do { \
 	int _plen; \
-	CBPRINTF_STATIC_PACKAGE(NULL, 0, _plen, Z_LOG_MSG2_ALIGN_OFFSET, \
-				CBPRINTF_PACKAGE_FIRST_RO_STR_CNT(_cstr_cnt), \
+	CBPRINTF_STATIC_PACKAGE(NULL, 0, _plen, Z_LOG_MSG2_ALIGN_OFFSET, flags, \
```
[the remark this answers is not in the document]
Comment (dcpleung) [reply include/logging/log_msg2.h]: weird... ran through twister without any issue...

**Comment by nordic-krch** [comment]
> Therefore, each string argument needs to be replaced with the full string.

@dcpleung does it mean that strings are inserted as value into the stream and not appended like in cbprintf package? Is this somehow standardized or would it be possible to somehow send package as is (converted to mipi syst stream)? Current approach requires quite a lot of processing. Also when package is statically created then ro/rw indexes will be appended for char pointers even when they were used for `%p`. Will it not disrupt the processing?

One more thing. I didn't see `source` being sent, any plans to support it?

**Comment by nordic-krch** [comment]
right, I only got to mipi STP specification and couldn't find anything there. Seems that i need to get into mipi_syst specification then.

**Comment by dcpleung** [comment]
Rebased after #43520 was merged.

**Review by MaureenHelm — DISMISSED** [review]

**Comment by dcpleung** [comment]
Added a README for the syst sample to show how to get back readable log messages.

**Review by nashif — APPROVED** [review]

**Review by nordic-krch — APPROVED** [review]
One minor comment.

I wonder if it wouldn't be better to just dump log message as is (similar to dictionary backend) or put it into some custom mipi STP stream and then have some python script that converts this stream to mipi syst format. That would be simpler on device side, handle `%p` and ro strings could also be sent by address and fetched on the host.

**Comment by MaureenHelm** [comment]
> Added a README for the syst sample to show how to get back readable log messages.

Thanks @dcpleung

File: subsys/logging/log_output_syst.c [hunk 3]
```diff
@@ -785,6 +784,394 @@ static int mipi_formatter(cbprintf_cb out, void *ctx,
...
+ * know which argument is integer, a 32-bit integer cannot be
+ * shrunk to 4 bytes for output. So error out here to prevent it
+ * being used at all.
+ */
+#ifdef CONFIG_64BIT
+#error "64-bit is not support at the moment!"
+#endif
+
+#ifdef CONFIG_64BIT
+#define MIPI_SYST_CATMSG_ARGS_COPY MIPI_SYST_CATALOG64_ARGS_COPY
```
[the remark this answers is not in the document]
Comment (dcpleung) [reply subsys/logging/log_output_syst.c:831]: Hopefully we can support 64-bit in the near future.

**Comment by dcpleung** [comment]
> I wonder if it wouldn't be better to just dump log message as is (similar to dictionary backend) or put it into some custom mipi STP stream and then have some python script that converts this stream to mipi syst format. That would be simpler on device side, handle `%p` and ro strings could also be sent by address and fetched on the host.

This is probably going to be problematic as people won't be able to use arbitrary MIPI tool to parse the log messages. Having a standard spec means multiple tools can be used (as drop-in replacement). Requiring an additional parsing steps breaks this model. I don't think people needing log messages through MIPI would want to deviate from standard.

**Pull request merged**